[clang-tidy] Create bugprone-incorrect-enable-shared-from-this check

MichelleCDjunaidi opened the pull request:
This checks that classes/structs inheriting from ``std::enable_shared_from_this`` does so with public inheritance, so it prevents crashes due to ``std::make_shared`` and ``shared_from_this()`` getting called when the internal weak pointer was not initialized (e.g. due to private inheritance).

MichelleCDjunaidi committed:
[clang-tidy] Add bugprone-public-enable-shared-from-this check

This check identifies classes deriving from std::enable_shared_from_this that does not inherit with the public keyword,
which may cause problems when std::make_shared is called for that class.

MichelleCDjunaidi committed:
Fix test to be compatible with llvm-lit, add docs

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/BugproneTidyModule.cpp:
```diff
@@ -139,6 +140,8 @@ class BugproneModule : public ClangTidyModule {
     CheckFactories.registerCheck<IncorrectEnableIfCheck>(
         "bugprone-incorrect-enable-if");
+    CheckFactories.registerCheck<PublicEnableSharedFromThisCheck>(
+        "bugprone-public-enable-shared-from-this");
```
name should be more a "non-public", as check should tell whats wrong.
i would call this check bugprone-incorrect-enable-shared-from-this, and then you could also check other constructions in future like:
```
class Bad2Example : enable_shared_from_this<BadExample> {};
```

PiotrZSL replied:
Or just call it bugprone-non-public-enable-shared-from-this

MichelleCDjunaidi replied:
Changed to bugprone-incorrect-enable-shared-from-this to keep the possibility of checking other constructions in the future open

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,45 @@
+
+  void PublicEnableSharedFromThisCheck::registerMatchers(MatchFinder *match_finder) {
+      match_finder->addMatcher(
+              cxxRecordDecl(
```
consider excluding system headers, to speed things up

MichelleCDjunaidi replied:
Oh, I didn't know that's possible in registerMatchers. Mind pointing me to a resource so I can figure that out?

PiotrZSL replied:
`unless(isExpansionInSystemHeader())`

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,45 @@
+  void PublicEnableSharedFromThisCheck::registerMatchers(MatchFinder *match_finder) {
+      match_finder->addMatcher(
+              cxxRecordDecl(
+                  hasAnyBase(
```
`hasAnyBase(cxxBaseSpecifier(...).bind("base"))`, check if this works, maybe with that you could avoid having loop in a check method

MichelleCDjunaidi replied:
![image](https://github.com/user-attachments/assets/e5c1716a-ad63-441c-959a-f9a5d6717285)
I've checked, the binding comes up empty handed (see that there's no hello2 output in the terminal)

Unless you have a better idea to make it not come up empty?

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,45 @@
+          const auto *BaseType = Base.getType()->getAsCXXRecordDecl();
+          if (BaseType && BaseType->getQualifiedNameAsString() == "std::enable_shared_from_this") {
+              SourceLocation InsertLoc = Base.getBeginLoc();
+              FixItHint Hint = FixItHint::CreateInsertion(InsertLoc, "public ");
```
what if it already got explicit "private", then it will end up with "private public" and thats invalid

MichelleCDjunaidi replied:
Fixed

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,45 @@
+          if (BaseType && BaseType->getQualifiedNameAsString() == "std::enable_shared_from_this") {
+              SourceLocation InsertLoc = Base.getBeginLoc();
+              FixItHint Hint = FixItHint::CreateInsertion(InsertLoc, "public ");
+              diag(EnableSharedClassDecl->getLocation(), "class %0 is not public even though it's derived from std::enable_shared_from_this", DiagnosticIDs::Warning)
```
this warning message doesn't make sense. there is no such thing like public/private class. This public/private inheritance.

MichelleCDjunaidi replied:
Thanks for correcting me, will get on it

MichelleCDjunaidi replied:
Warning is changed. Matcher changes/loop removal if possible are still WIP

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,45 @@
+              SourceLocation InsertLoc = Base.getBeginLoc();
+              FixItHint Hint = FixItHint::CreateInsertion(InsertLoc, "public ");
+              diag(EnableSharedClassDecl->getLocation(), "class %0 is not public even though it's derived from std::enable_shared_from_this", DiagnosticIDs::Warning)
+              << EnableSharedClassDecl->getNameAsString()
```
no need for getNameAsString

MichelleCDjunaidi replied:
Fixed

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.h:
```diff
@@ -0,0 +1,33 @@
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+  bool isLanguageVersionSupported(const LangOptions &LangOpts) const override {
+    return LangOpts.CPlusPlus;
```
CPlusPlus11

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -98,6 +98,11 @@ Improvements to clang-tidy
+- New :doc:`bugprone-public-enable-shared-from-this
+  <clang-tidy/checks/bugprone/public-enable-shared-from-this>` check.
+
+  Checks if a class deriving from enable_shared_from_this has public access specifiers, because otherwise the program will crash when shared_from_this is called.
```
again, wrong. "Check if class publicly derive from std::enable_shared_from_this ..."
And wrap on 80 column.

MichelleCDjunaidi replied:
Right, will run git clang format, thanks for the reminder and feedback

MichelleCDjunaidi replied:
ok manually reformatted since git clang format doesn't work on rst, and changed documentation

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/public-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,26 @@
+                b_ex->bar();
+        }
+
+When ``using_not_public()`` is called, this code will crash.
```
no it won't crash, it will thrown `std::bad_weak_ptr`, if application didin't implement proper exception handling then it would crash.

MichelleCDjunaidi replied:
Yeah, will fix the desc

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/public-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,45 @@
+namespace std {
+
+    template <typename T> class enable_shared_from_this {};
+
```
std namespace should end here

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/public-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,45 @@
+    class Bad2Example : std::enable_shared_from_this<Bad2Example> {};
+    // CHECK-MESSAGES: :[[@LINE-1]]:7: warning: class Bad2Example is not public even though it's derived from std::enable_shared_from_this [bugprone-public-enable-shared-from-this]
+    // CHECK-FIXES: public std::enable_shared_from_this<Bad2Example>
+
```
add test:
class Bad3Example : private std::enable_shared_from_this<Bad3Example> {};

MichelleCDjunaidi replied:
Done

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/public-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,45 @@
+    class GoodExample : public enable_shared_from_this<GoodExample> {
+    };
+
+    struct BaseClass {
+
+        void print() {
+            (void) State;
+            (void) Requester;
+        }
+        bool State;
+        int Requester;
+    };
+
+    class InheritPrivateBaseClass : BaseClass {
+        public:
+            void additionalFunction() {
+                (void) ID;
+            }
+        private: 
+            int ID;
+    };
+
+    class InheritPublicBaseClass : public BaseClass {
+        public:
+            void additionalFunction() {
+                (void) ID;
+            }
+        private: 
+            int ID;
+    };
```
Those classes does not add anything, so maybe remove them, or you missing something here

MichelleCDjunaidi replied:
Ah, so I don't need to demonstrate that it doesn't match things with public inheritance? Understood, will do

MichelleCDjunaidi replied:
Removed

PiotrZSL reviewed (changes requested):
I'm fine with idea behind this check. Just few fixes are needed.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,45 @@
+//===--- PublicEnableSharedFromThisCheck.cpp - clang-tidy -------------------------===//
```
Please follow 80 characters limit.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/PublicEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,45 @@
+          result.Nodes.getNodeAs<CXXRecordDecl>("not-public-enable-shared");
+
+      for (const auto &Base : EnableSharedClassDecl->bases()) {
+          const auto *BaseType = Base.getType()->getAsCXXRecordDecl();
```
Please don't use `auto` unless type is explicitly stated in same statement or iterator.

MichelleCDjunaidi replied:
Fixed

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -98,6 +98,11 @@ Improvements to clang-tidy
+- New :doc:`bugprone-public-enable-shared-from-this
+  <clang-tidy/checks/bugprone/public-enable-shared-from-this>` check.
+
+  Checks if a class deriving from enable_shared_from_this has public access specifiers, because otherwise the program will crash when shared_from_this is called.
```
80 characters limit. Please highlight language constructs with double back-ticks.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/public-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,26 @@
+bugprone-public-enable-shared-from-this
+=======================================
+
+Checks that classes/structs inheriting from ``std::enable_shared_from_this`` derive it with the ``public`` access specifier. If not, then issue a FixItHint that can be applied.
```
80 characters limit.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/public-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,26 @@
+
+Consider the following code:
+.. code-block:: c++
+        #include <memory>
```
Indentation should be reduced.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/public-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,26 @@
+Checks that classes/structs inheriting from ``std::enable_shared_from_this`` derive it with the ``public`` access specifier. If not, then issue a FixItHint that can be applied.
+
+Consider the following code:
+.. code-block:: c++
```
Please separate with newlines.

EugeneZelenko commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/public-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,45 @@
+
+namespace std {
+
+    template <typename T> class enable_shared_from_this {};
```
Please format with Clang-format.

MichelleCDjunaidi replied:
Manually formatted.

njames93 reviewed (commented):
How are classes that derive other classes that implement `enable_shared_from_this`

```cpp
class A : public std::enable_shared_from_this {};
class B : private A{};
```

MichelleCDjunaidi committed:
Change name to bugprone-incorrect-enable-shared-from-this.

WIP in regards to test file and matchers, do not re-review yet.

MichelleCDjunaidi committed:
Fix ``FixItHint`` of public inheritance, implement ``std::`` check

Note: Maybe should be separate checks

MichelleCDjunaidi committed:
Stated type instead of using auto in the ``check`` method

MichelleCDjunaidi commented:
> How are classes that derive other classes that implement `enable_shared_from_this`
> 
> ```c++
> class A : public std::enable_shared_from_this {};
> class B : private A{};
> ```

Looking at the AST dump that clang-check gives, it seems that B doesn't have any knowledge of A's internals, so the current approach will not work. Correct me if I'm wrong though

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,96 @@
+//===--- IncorrectEnableSharedFromThisCheck.cpp - clang-tidy
+//-------------------------===//
```
Please merge to previous line. See other checks as example.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,96 @@
+      if (BaseType && BaseType->getQualifiedNameAsString() ==
+                          "std::enable_shared_from_this") {
+        SourceRange ReplacementRange = Base.getSourceRange();
+        std::string ReplacementString =
```
```suggestion
        const std::string ReplacementString =
```

Will be good idea to run `misc-const-correctness` to find other variables that should be `const`.

MichelleCDjunaidi replied:
Ran and fixed

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,96 @@
+    }
+  }
+}
+} // namespace clang::tidy::bugprone
```
Please separate with newline.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,34 @@
+
+Check if class/struct publicly derives from ``std::enable_shared_from_this``,
+because otherwise when ``shared_from_this`` is called it will throw 
+``std::bad_weak_ptr``. Issues a ``FixItHint`` that can be applied.
```
Last statement is not needed, this information is provided in list of checks.

EugeneZelenko commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/incorrect-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,57 @@
+//FIXME: can't do anything about this, clang-check -ast-dump doesn't show A's internals in class B's AST
+class A : public std::enable_shared_from_this<A> {};
+class B : private A{};
+
```
Excessive newline.

njames93 commented:
> > How are classes that derive other classes that implement `enable_shared_from_this`
> > 
> > ```c++
> > class A : public std::enable_shared_from_this {};
> > class B : private A{};
> > ```
> 
> Looking at the AST dump that clang-check gives, it seems that B doesn't have any knowledge of A's internals, so the current approach will not work. Correct me if I'm wrong though

I think you'd have to recursively check the inheritance of a class and if you have any private that derives from `std::enable_shared_from_this`

PiotrZSL commented:
Yu could use isDerivedFrom, should do a trick

njames93 commented:
Actually an ASTVisitor approach could be better here, this way you could create a set of all the `CXXRecordDecls` that publicly inherit from or transitively inherit from `::std::enable_shared_from_this` during traversal

MichelleCDjunaidi commented:
> Actually an ASTVisitor approach could be better here, this way you could create a set of all the `CXXRecordDecls` that publicly inherit from or transitively inherit from `::std::enable_shared_from_this` during traversal

To confirm, @njames93 do you mean clang::RecursiveASTVisitor? What I'm getting from this conversation is maintaining the name of the matched classes in a vector and checking for every other class if it inherits from `::std::enable_shared_from_this` or classes in the vector.

MichelleCDjunaidi committed:
Refactored const variables and cleanup newlines/formatting

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,34 @@
+    #include <memory>
+
+    class BadExample : std::enable_shared_from_this<BadExample> {
+    // warning: inheritance from std::enable_shared_from_this 
+    // should be public inheritance,
+    // otherwise the internal weak_ptr won't be initialized 
+    // [bugprone-incorrect-enable-shared-from-this]
```
The diagnostics of a check are normally not copy-pasted to the documentation. Maybe: 
> privately inheriting from `std::enable_shared_from_this` results in an uninitialized `weak_ptr` when calling `shared_from_this`

Or add smaller comments:

```
 #include <memory>

    // private inheritance
    class BadExample : std::enable_shared_from_this<BadExample> {
        public:

        // `shared_from_this` returns uninitialized `weak_ptr`
        BadExample* foo() { return shared_from_this().get(); }
        void bar() { return; }
    };

    void using_not_public() {
        auto bad_example = std::make_shared<BadExample>();
        auto* b_ex = bad_example->foo();
        b_ex->bar();
    }
```

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/incorrect-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,56 @@
+
+//FIXME: can't do anything about this, clang-check -ast-dump doesn't show A's internals in class B's AST
+class A : public std::enable_shared_from_this<A> {};
+class B : private A{};
```
Missing newline

5chmidti reviewed (changes requested):
> maintaining the name of the matched classes in a vector and checking for every other class if it inherits from `::std::enable_shared_from_this` or classes in the vector.

You wouldn't want to save the name, because then you would do a lot of string comparisons when you could do pointer comparisons (which would require using `RDecl->getCanonical()` to ensure you get the same record decl, I believe).
Instead of a vector, you could use a set. Then checking the base classes would simply be a call to `contains` (`llvm::SmallPtrSet`?).

> Wouldn't this necessitate changing the matchers and the check to instead allow RecursiveASTVisitor to handle the context instead?

Kind of. In this case, you would match the translation unit and call `TraverseTranslationUnitDecl` on it.
Then you implement `bool VisitCXXRecordDecl(CXXRecordDecl* RDecl)` and check if `RDecl` is of type `std::enable_shared_from_this` and adding that to the set if it is. Otherwise, checking if it publicly inherits from any class that is contained in the set, and adding it to the set if it is. If the inheritance is not public, then you can issue a diagnostic. You can save a pointer to the check (`ClangTidyCheck* Check`) to do that (`Check->diag(...)`).

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,34 @@
+.. title:: clang-tidy - bugprone-incorrect-enable-shared-from-this
+
+bugprone-incorrect-enable-shared-from-this
+=======================================
```
Please make same length as check name.

MichelleCDjunaidi committed:
Fix documentation, comments, newlines, sort module registration properly

MichelleCDjunaidi commented:
> > maintaining the name of the matched classes in a vector and checking for every other class if it inherits from `::std::enable_shared_from_this` or classes in the vector.
> 
> You wouldn't want to save the name, because then you would do a lot of string comparisons when you could do pointer comparisons (which would require using `RDecl->getCanonical()` to ensure you get the same record decl, I believe). Instead of a vector, you could use a set. Then checking the base classes would simply be a call to `contains` (`llvm::SmallPtrSet`?).
> 
> > Wouldn't this necessitate changing the matchers and the check to instead allow RecursiveASTVisitor to handle the context instead?
> 
> Kind of. In this case, you would match the translation unit and call `TraverseTranslationUnitDecl` on it. Then you implement `bool VisitCXXRecordDecl(CXXRecordDecl* RDecl)` and check if `RDecl` is of type `std::enable_shared_from_this` and adding that to the set if it is. Otherwise, checking if it publicly inherits from any class that is contained in the set, and adding it to the set if it is. If the inheritance is not public, then you can issue a diagnostic. You can save a pointer to the check (`ClangTidyCheck* Check`) to do that (`Check->diag(...)`).

Thanks for the pointers! It's really helpful to have a starting point to learn the best practices, I'll try to figure this out and report back.

MichelleCDjunaidi commented:
@5chmidti so something like this then? ``AssignmentInIfConditionCheck`` in bugprone nests the ``RecursiveASTVisitor``, so I tried following suit, but this feels unreadable and I'll refactor it once it's working. Right now this first draft isn't outputting anything and so fails the llvm-lit test (will troubleshoot and commit once it's working; if anyone has any idea why this doesn't work, do tell).
```
//===--- IncorrectEnableSharedFromThisCheck.cpp - clang-tidy --------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//

#include "IncorrectEnableSharedFromThisCheck.h"
#include "clang/AST/DeclCXX.h"
#include "clang/AST/ASTContext.h"
#include "clang/AST/RecursiveASTVisitor.h"
#include "clang/ASTMatchers/ASTMatchFinder.h"
#include "clang/Basic/Specifiers.h"
#include "llvm/ADT/SmallPtrSet.h"

using namespace clang::ast_matchers;

namespace clang::tidy::bugprone {

void IncorrectEnableSharedFromThisCheck::registerMatchers(
    MatchFinder *Finder) {
  Finder->addMatcher(translationUnitDecl(), this);
}

void IncorrectEnableSharedFromThisCheck::check(
  const MatchFinder::MatchResult &Result) {

  class Visitor : public RecursiveASTVisitor<Visitor> {
    IncorrectEnableSharedFromThisCheck &Check;
    llvm::SmallPtrSet<CXXRecordDecl*, 8> *EnableSharedClassSet;

  public:
    explicit Visitor(IncorrectEnableSharedFromThisCheck &Check) : Check(Check), EnableSharedClassSet(EnableSharedClassSet) {}
    bool VisitCXXRecordDecl(CXXRecordDecl *RDecl) {

      class BaseVisitor : public RecursiveASTVisitor<BaseVisitor> {
        IncorrectEnableSharedFromThisCheck &Check;
        llvm::SmallPtrSet<CXXRecordDecl*, 8> *EnableSharedClassSet;
        CXXRecordDecl *RDecl;

      public:
        explicit BaseVisitor(IncorrectEnableSharedFromThisCheck &Check) : 
        Check(Check), EnableSharedClassSet(EnableSharedClassSet), RDecl(RDecl) {}

        bool VisitCXXBaseSpecifier(CXXBaseSpecifier &Base) {
          const CXXRecordDecl *BaseType = Base.getType()->getAsCXXRecordDecl();
          const clang::AccessSpecifier AccessSpec = Base.getAccessSpecifier();

          if ( BaseType && BaseType->getQualifiedNameAsString() ==
                              "enable_shared_from_this") {
            if (AccessSpec == clang::AS_public) {
              const SourceLocation InsertLocation = Base.getBaseTypeLoc();
              const FixItHint Hint = FixItHint::CreateInsertion(InsertLocation, "std::");
              Check.diag(RDecl->getLocation(),
                  "Should be std::enable_shared_from_this", DiagnosticIDs::Warning)
                  << Hint;
              
            } else {
              const SourceRange ReplacementRange = Base.getSourceRange();
              const std::string ReplacementString =
                  "public std::" + Base.getType().getAsString();
              const FixItHint Hint =
                  FixItHint::CreateReplacement(ReplacementRange, ReplacementString);
              Check.diag(RDecl->getLocation(),
                  "Should be std::enable_shared_from_this and "
                  "inheritance from std::enable_shared_from_this should be public "
                  "inheritance, otherwise the internal weak_ptr won't be "
                  "initialized",
                  DiagnosticIDs::Warning)
                  << Hint;
            }
          }

          if ( BaseType && BaseType->getQualifiedNameAsString() ==
                              "std::enable_shared_from_this") {
            CXXRecordDecl *CanonicalRDecl = RDecl->getCanonicalDecl();
            EnableSharedClassSet->insert(CanonicalRDecl);
            if (AccessSpec != clang::AS_public) {
              const SourceRange ReplacementRange = Base.getSourceRange();
              const std::string ReplacementString =
                  "public " + Base.getType().getAsString();
              const FixItHint Hint =
                  FixItHint::CreateReplacement(ReplacementRange, ReplacementString);
              Check.diag(
                  RDecl->getLocation(),
                  "inheritance from std::enable_shared_from_this should be public "
                  "inheritance, otherwise the internal weak_ptr won't be initialized",
                  DiagnosticIDs::Warning)
                  << Hint;
            }
          } 
          else if (EnableSharedClassSet->contains(BaseType)) {
            if (AccessSpec != clang::AS_public) {
              const SourceRange ReplacementRange = Base.getSourceRange();
              const std::string ReplacementString =
                  "public " + Base.getType().getAsString();
              const FixItHint Hint =
                  FixItHint::CreateReplacement(ReplacementRange, ReplacementString);
              Check.diag(
                  RDecl->getLocation(),
                  "inheritance from std::enable_shared_from_this should be public "
                  "inheritance, otherwise the internal weak_ptr won't be initialized",
                  DiagnosticIDs::Warning)
                  << Hint;
            }            
          }
          return true;
        }
      };
      
      BaseVisitor(Check).TraverseCXXRecordDecl(RDecl);
      return true;
    }

  };
 
  Visitor(*this).TraverseAST(*Result.Context);
  
}

} // namespace clang::tidy::bugprone
```

MichelleCDjunaidi committed:
Switch implementation to RecursiveASTVisitor

MichelleCDjunaidi committed:
Fix test case formatting

MichelleCDjunaidi committed:
Fix check code formatting to follow LLVM style

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+              RDecl->getLocation(),
+              "Should be std::enable_shared_from_this and "
+              "inheritance from std::enable_shared_from_this should be public "
+              "inheritance, otherwise the internal weak_ptr won't be "
```
noneone care/know about "internal weak_ptr, write how it would behave from user perspective, start with lower caps.
For me this is too long. it should say: whats wrong (+why it's wrong), how to fix (+why fix)

MichelleCDjunaidi replied:
changed to "this is not publicly inheriting from std::enable_shared_from_this, will cause unintended behaviour on shared_from_this. fix this by making it public inheritance".

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+
+    bool VisitCXXRecordDecl(CXXRecordDecl *RDecl) {
+      for (const auto &Base : RDecl->bases()) {
+        VisitCXXBaseSpecifier(Base, RDecl);
```
shouldn't this be called anyway ? Maybe try with traverse instead of visit

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+      }
+      for (const auto &Base : RDecl->bases()) {
+        const CXXRecordDecl *BaseType = Base.getType()->getAsCXXRecordDecl();
+        if (BaseType && BaseType->getQualifiedNameAsString() ==
```
that getQualifiedNameAsStringis ugly, and did you check if getAsCXXRecordDecl will work with typedefs ?

MichelleCDjunaidi replied:
what would be a good alternative then to check the class? I'll check if it works with typedefs

MichelleCDjunaidi replied:
checked, it works with typedef. will update implementation to use [isInStdNamespace](https://clang.llvm.org/doxygen/classclang_1_1Decl.html#a066b012f94431b5bba21d19715a274f4) and getName as per comment https://github.com/llvm/llvm-project/pull/102299#discussion_r1719049161

MichelleCDjunaidi replied:
updated.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+        const CXXRecordDecl *BaseType = Base.getType()->getAsCXXRecordDecl();
+        if (BaseType && BaseType->getQualifiedNameAsString() ==
+                            "std::enable_shared_from_this") {
+          EnableSharedClassSet.insert(RDecl->getCanonicalDecl());
```
instead of getCanonicalDecl would be better to take definition, as getCanonicalDecl is a first declaration (if i remember correctly)

MichelleCDjunaidi replied:
https://github.com/llvm/llvm-project/pull/102299#pullrequestreview-2232523871

I was suggested to use getCanonicalDecl, would getDefinition result in the same pointer?

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+        : Check(Check) {}
+
+    bool VisitCXXRecordDecl(CXXRecordDecl *RDecl) {
+      for (const auto &Base : RDecl->bases()) {
```
exclude non-definitions here quickly to avoid processing same class multiple times.

MichelleCDjunaidi replied:
will do.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+      const clang::AccessSpecifier AccessSpec = Base.getAccessSpecifier();
+
+      if (BaseType &&
+          BaseType->getQualifiedNameAsString() == "enable_shared_from_this") {
```
why no `::std::``

MichelleCDjunaidi replied:
this was to check for missing std::

MichelleCDjunaidi replied:
this part was removed in accordance to https://github.com/llvm/llvm-project/pull/102299#discussion_r1719046306

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+          BaseType->getQualifiedNameAsString() == "enable_shared_from_this") {
+        if (AccessSpec == clang::AS_public) {
+          const SourceLocation InsertLocation = Base.getBaseTypeLoc();
+          const FixItHint Hint = FixItHint::CreateInsertion(InsertLocation, "std::");
```
wrong, check if current class is std::enable_shared_from_this, someone may be using "using namespace std", and i don't see an issue with missing "std::", that should be separate generic check, and I think that there is being one created (or someone were working on it)

MichelleCDjunaidi replied:
ah, I implemented this missing std:: thing because of this suggestion. I didn't know there was a generic check for it.

https://github.com/llvm/llvm-project/pull/102299#discussion_r1706831159

so, do I just remove this part of the check?

MichelleCDjunaidi replied:
removed std::check.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+        } else {
+          const SourceRange ReplacementRange = Base.getSourceRange();
+          const std::string ReplacementString =
+              "public std::" + Base.getType().getAsString();
```
doesn't make sense, why not write then `public std::enable_shared_from_this`

MichelleCDjunaidi replied:
the getType().getAsString is already including enable_shared_from_this<ClassName>, so doing "public std::enable_shared_from_this" + Base.getType().getAsString()  would result in public std::enable_shared_from_thisenable_shared_from_this<ClassName>.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,120 @@
+      return true;
+    }
+
+    bool VisitCXXBaseSpecifier(const CXXBaseSpecifier &Base,
```
this entire function is too complicated to what it does, for example you got multiple compares to std::enable_shared_from_this, or :enable_shared_from_this, that should be eliminated. 
And instead of using getQualifiedNameAsString, [isInStdNamespace](https://clang.llvm.org/doxygen/classclang_1_1Decl.html#a066b012f94431b5bba21d19715a274f4) and getName should be used.

PiotrZSL reviewed (changes requested):
for me oryginal version that used registerMatchers instead of visitor were way better, specially that there are matchers that could be used for this, or could be written.
And even for performance reasons it could be implemented in slightly better way by simply very quickly exclude all classes that do not derive from std::enable_shared_from_this.

If `boost::enable_shared_from_this` got same flaw, then this check should also support it. And this mean it maybe should be configurable.

Edit: I understand why @njames93 mentioned Visitor, but thing is that in most projects not many classes actually used `std::enable_shared_from_this` and there will be no visible performance gain from "remembering" list of classes, and as we could see this approach makes this check more complicated.

MichelleCDjunaidi committed:
Optimize traversal, add alias and typedef testing, remove missing std:: check

In accordance to https://github.com/llvm/llvm-project/pull/102299#discussion_r1719046306, the
feature that was suggested in https://github.com/llvm/llvm-project/pull/102299#discussion_r1706831159
has been removed.

MichelleCDjunaidi commented:
@5chmidti @njames93 @PiotrZSL Thoughts regarding which direction to take?

5chmidti commented:
> for me oryginal version that used registerMatchers instead of visitor were way better, specially that there are matchers that could be used for this, or could be written.

We do have the `isDerivedFrom` matcher, actually.

> in most projects not many classes actually used std::enable_shared_from_this and there will be no visible performance gain from "remembering" list of classes, and as we could see this approach makes this check more complicated.

Maybe we should just go with what is simpler, and check the performance of the implementation, but I think it could be more performant.

```c++
struct A {};
struct B : public A {};
struct C : public B {};
struct D : public B {};
```

When we are visiting a definition of a class, we know that every base class has been visited already, so we do not need to traverse the inheritance hierarchy and instead just need to check our set.
We mark `A` as either an `enable_shared_from_this` class, or not.
We check for `B`, if any of its base classes is already in the set, and the same goes for `C` and `D`. The base classes of `C` and `D` are never actually transitively traverse when we check `C` and `D`, but instead we do a lookup because we already computed the answer.

---

@MichelleCDjunaidi 

Untested, but this should be roughly what needs to be done (minus the check for a missing `std::` in front of `enable_shared_from_this`).

```c++
  bool VisitCXXRecordDecl(CXXRecordDecl *RDecl) {
      if (RDecl->getQualifiedNameAsString() ==
                            "std::enable_shared_from_this")
          EnableSharedClassSet.insert(RDecl->getCanonicalDecl());

      for (const auto &Base : RDecl->bases()) {
        const auto* BaseRecord = Base.getType()->getAsCXXRecordDecl()->getCanonicalDecl();
        
        if (EnableSharedClassSet.contains(BaseRecord))
        {
            if (Base.getAccessSpecifier() != AccessSpecifier::AS_public)
               Check.diag(...);
            else
                EnableSharedClassSet.insert(RDecl->getCanonicalDecl());
        }
      }
      return true;
    }
```

Your `VisitCXXBaseSpecifier` has the same name as the other visitor functions, but it is technically not in the format of the `RecursiveASTVisitor` and may be confusing. `Handle...` or `Check...` would be better IMO.

Either way, the direction is still an open question

MichelleCDjunaidi committed:
refactor: simplify to VisitCXXRecordDecl only instead of TraverseCXXRecordDecl, VisitCXXRecordDecl, VisitCXXBaseSpecifier

Logic remains unchanged.

MichelleCDjunaidi committed:
refactor: exclude non-definitions from repeated processing

MichelleCDjunaidi committed:
more detailed diag for classes with bases from EnableSharedClassSet

MichelleCDjunaidi commented:
@5chmidti Thanks for the pointers, I've implemented your suggestion.

For future knowledge, what do you mean by 
> it is technically not in the format of the RecursiveASTVisitor and may be confusing
? Is there a guide that can point out what format Visit should have?

MichelleCDjunaidi commented:
@5chmidti @njames93 @PiotrZSL Any updated feedback on this?

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,105 @@
+        const auto *BaseRecord =
+            Base.getType()->getAsCXXRecordDecl()->getCanonicalDecl();
+
+        if (EnableSharedClassSet.contains(BaseRecord) ||
+            isStdEnableSharedFromThis(BaseRecord)) {
+
+          if (isStdEnableSharedFromThis(BaseRecord) &&
+              Base.getAccessSpecifier() != clang::AS_public) {
```
You are calling `isStdEnableSharedFromThis(BaseRecord)` twice with the same value. Please save the result before the `if` and reuse the computed value (yes, it's a cheap computation).

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,105 @@
+                << Hint;
+          } else if (EnableSharedClassSet.contains(BaseRecord) &&
+                     Base.getAccessSpecifier() != clang::AS_public) {
+            // actually the same FixItHint, just different diag
```
Then you could use the `%select` operator of the diagnostic builder to merge both cases:

```c++
diag(RDecl->getLocation(), 
     "%select{'std::enable_shared_from_this'|%1}0") << isStdEnableSharedFromThis(BaseRecord) << BaseRecord;
```
Note the `0` and `1` are indices to the `<<`-ed expressions.

MichelleCDjunaidi replied:
wonderful, didn't know this one existed. Glad to be able to get rid of the horrible code duplication/if nesting.

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,33 @@
+        b_ex->bar();
+    }
+
+Using ``libstd`` implementation, `shared_from_this`` will throw 
```
`Using the ..`

`shared_from_this` is missing one leading backtick

5chmidti reviewed (commented):
> Side note, for future knowledge, what do you mean by
> 
> > it is technically not in the format of the RecursiveASTVisitor and may be confusing
> 
> ? Is there a guide that can point out what format Visit... should have?

My bad, it actually was named correct. But, the format is: `Traverse<Name>`, `WalkUpFrom<Name>` and `Visit<Name>`, where `<Name>` is replaced with AST nodes.
https://clang.llvm.org/doxygen/classclang_1_1RecursiveASTVisitor.html

However, https://clang.llvm.org/doxygen/classclang_1_1RecursiveASTVisitor.html says that
> These three method groups are tiered (Traverse* > WalkUpFrom* > Visit*). A method (e.g. Traverse*) may call methods from the same tier (e.g. other Traverse*) or one tier lower (e.g. WalkUpFrom*).

So, calling `Visit` from `Traverse` may result in some issues.

> Any updated feedback on this?

I reviewed the code again, but which of the two approaches to take is still in question. (ping for reviewers)

MichelleCDjunaidi committed:
refactor to use diag %select, fix docs typo

MichelleCDjunaidi committed:
add space to code block in docs

MichelleCDjunaidi committed:
update test CHECK-MESSAGES to check for class name

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+      for (const auto &Base : RDecl->bases()) {
+        const auto *BaseRecord =
+            Base.getType()->getAsCXXRecordDecl()->getCanonicalDecl();
+        const auto isStdEnableSharedFromThisBool =
+            isStdEnableSharedFromThis(BaseRecord);
```
Variables are `CamelCase` in LLVM. A better name could be `IsBaseStdEnableSharedFromThis`.

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+        : Check(Check) {}
+
+    bool VisitCXXRecordDecl(CXXRecordDecl *RDecl) {
+      if (!RDecl->hasDefinition()) {
+        return true;
+      }
```
Nit: control flow with single statements do not use compound statements in LLVM

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+                ReplacementRange, ReplacementString);
+            Check.diag(RDecl->getLocation(),
+                       "%2 is not publicly inheriting from "
+                       "%select{%1|'std::enable_shared_from_this',}0 "
+                       "%select{which inherits from "
+                       "'std::enable_shared_from_this', |}0 "
```
These can be merged, because they have the same condition

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+                       "%select{which inherits from "
+                       "'std::enable_shared_from_this', |}0 "
+                       "will cause unintended behaviour "
+                       "on 'shared_from_this'. fix this by making it public "
```
WDYT?: `on` -> `when using` or `[when] calling`

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+                       "%select{which inherits from "
+                       "'std::enable_shared_from_this', |}0 "
+                       "will cause unintended behaviour "
+                       "on 'shared_from_this'. fix this by making it public "
+                       "inheritance",
```
I have only seen two parts of a diagnostic be separated by a `;`, instead of using full stops with sentences `.`.

E.g.,: `... 'shared_from_this'; make the inheritance public`

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace clang::tidy::bugprone {
+
+// Checks if class/struct publicly inherits from
+// ``std::enable_shared_from_this``, because otherwise when ``shared_from_this``
+// is called unintended behaviour will occur
```
nit: `Checks if a class or struct` (including both of the other occurences of this sentence)

nit (not entirely sure about these):
> ..., because unintended behavior will otherwise occur when calling ``shared_from_this``.

> ..., because the inheritance will otherwise not be considered when calling ``shared_from_this``.

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/incorrect-enable-shared-from-this.cpp:
Please add tests with macros involved.

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/incorrect-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,121 @@
+class dummy_class2 {};
+
+class BadMultiClassExample : std::enable_shared_from_this<BadMultiClassExample>, dummy_class1 {};
+// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: BadMultiClassExample is not publicly inheriting from 'std::enable_shared_from_this', will cause unintended behaviour on 'shared_from_this'. fix this by making it public inheritance [bugprone-incorrect-enable-shared-from-this]
```
The wording of this diagnostic case can be improved:
Either

> 'std::enable_shared_from_this', which will cause unintended

or

> 'BadMultiClassExample' not publicly inheriting

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+                       "inheritance",
+                       DiagnosticIDs::Warning)
+                << isStdEnableSharedFromThisBool
+                << BaseRecord->getNameAsString() << RDecl->getNameAsString()
```
Don't stream the names of these into diagnostics. Instead, you can simply stream the pointers, and they will automatically be converted to their name. These will also automatically be wrapped in `'`, which both `%1` and `%2` are currently missing.

MichelleCDjunaidi commented:
Thanks for the feedback! Currently working on it.

MichelleCDjunaidi committed:
[NFC] in progress change diag

MichelleCDjunaidi committed:
add macros to test file and revise CHECK-MESSAGES

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,92 @@
+        EnableSharedClassSet.insert(RDecl->getCanonicalDecl());
+
+      for (const auto &Base : RDecl->bases()) {
+        const auto *BaseRecord =
```
`auto` should not be used - types is not spelled explicitly.

MichelleCDjunaidi committed:
change auto vars to explicit typing

MichelleCDjunaidi commented:
@EugeneZelenko @5chmidti Any updates on the feedback? Is this ready to merge?

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -98,6 +98,13 @@ Improvements to clang-tidy
+- New :doc:`bugprone-incorrect-enable-shared-from-this
+  <clang-tidy/checks/bugprone/incorrect-enable-shared-from-this>` check.
+
+  Check if class/structs publicly inherits from ``std::enable_shared_from_this``,
```
```suggestion
  Detects if class/structs publicly inherits from ``std::enable_shared_from_this``,
```
This is continuation of previous statement, so `check checks` does not sound well.

MichelleCDjunaidi replied:
Noted, will do.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,34 @@
+        b_ex->bar();
+    }
+
+Using ``libstd`` implementation, ``shared_from_this`` will throw 
```
```suggestion
Using `libstd` implementation, ``shared_from_this`` will throw 
```

`libstdc++`?

MichelleCDjunaidi replied:
Yeah, libstdc++. Will change.

MichelleCDjunaidi committed:
update docs to fix wording

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,92 @@
+
+          if (Base.getAccessSpecifier() != clang::AS_public) {
+            const SourceRange ReplacementRange = Base.getSourceRange();
+            const std::string ReplacementString =
+                // Base.getType().getAsString() results in
+                // std::enable_shared_from_this<ClassName> or
+                // alias/typedefs of std::enable_shared_from_this<ClassName>
+                "public " + Base.getType().getAsString();
+            const FixItHint Hint = FixItHint::CreateReplacement(
+                ReplacementRange, ReplacementString);
+            Check.diag(RDecl->getLocation(),
+                       "%2 is not publicly inheriting from "
+                       "%select{%1 which inherits from "
+                       "'std::enable_shared_from_this',|'std::enable_shared_"
+                       "from_this',}0 "
+                       "which will cause unintended behaviour "
+                       "when using 'shared_from_this'; make the inheritance "
+                       "public",
+                       DiagnosticIDs::Warning)
+                << IsStdEnableSharedFromThisBool << BaseRecord << RDecl << Hint;
+          }
+
+          EnableSharedClassSet.insert(RDecl->getCanonicalDecl());
+        }
+      }
+      return true;
+    }
+
+  private:
+    // FIXME: configure this for boost in the future
+    bool isStdEnableSharedFromThis(const CXXRecordDecl *RDecl) {
+      // this is the equivalent of
+      // RDecl->getQualifiedNameAsString() == "std::enable_shared_from_this"
+      return RDecl->getName() == "enable_shared_from_this" &&
+             RDecl->isInStdNamespace();
+    }
+  };
+
+  Visitor(*this).TraverseAST(*Result.Context);
+}
```
So..., I've decided to go ahead and benchmark the RAV solution and the corresponding matcher based solution... and they perform the same XD

So you don't have to re-implement it as well, here is the implementation:

```suggestion
void IncorrectEnableSharedFromThisCheck::registerMatchers(MatchFinder *Finder) {
  const auto EnableSharedFromThis =
      cxxRecordDecl(hasName("enable_shared_from_this"), isInStdNamespace());
  const auto QType = hasCanonicalType(hasDeclaration(
      cxxRecordDecl(
          anyOf(EnableSharedFromThis.bind("enable_rec"),
                cxxRecordDecl(hasAnyBase(cxxBaseSpecifier(
                    isPublic(), hasType(hasCanonicalType(
                                    hasDeclaration(EnableSharedFromThis))))))))
          .bind("base_rec")));

  Finder->addMatcher(
      cxxRecordDecl(
          hasDirectBase(cxxBaseSpecifier(unless(isPublic()), hasType(QType))
                            .bind("base")))
          .bind("derived"),
      this);
}

void IncorrectEnableSharedFromThisCheck::check(
    const MatchFinder::MatchResult &Result) {
  const auto *BaseSpec = Result.Nodes.getNodeAs<CXXBaseSpecifier>("base");
  const auto *Base = Result.Nodes.getNodeAs<CXXRecordDecl>("base_rec");
  const auto *Derived = Result.Nodes.getNodeAs<CXXRecordDecl>("derived");
  const bool IsEnableSharedFromThisDirectBase =
      Result.Nodes.getNodeAs<CXXRecordDecl>("enable_rec") == Base;

  const SourceRange ReplacementRange = BaseSpec->getSourceRange();
  const std::string ReplacementString =
      // Base.getType().getAsString() results in
      // std::enable_shared_from_this<ClassName> or
      // alias/typedefs of std::enable_shared_from_this<ClassName>
      "public " + BaseSpec->getType().getAsString();
  const FixItHint Hint =
      FixItHint::CreateReplacement(ReplacementRange, ReplacementString);
  diag(Derived->getLocation(),
       "%2 is not publicly inheriting from "
       "%select{%1 which inherits from |}0'std::enable_shared_"
       "from_this', "
       "which will cause unintended behaviour "
       "when using 'shared_from_this'; make the inheritance "
       "public",
       DiagnosticIDs::Warning)
      << IsEnableSharedFromThisDirectBase << Base << Derived << Hint;
}
```

173 seconds for the `clang-tools-extra` folder, using this check (no issues detected, but the checking is done either way)

MichelleCDjunaidi replied:
alright, will revert back to matcher based check and remove unused headers.

5chmidti reviewed (commented):
This is 'basically' ready. Sorry for the delay.

---

As stated in the comment, I've benchmarked the RAV and matcher-based solution, and they perform the same. So the actual (performance) benefit of using the RAV does not actually exist. I'd suggest using the matcher based solution instead, as matchers should be the preferred tool in clang-tidy, and the implementation is much shorter/easier to understand at a glance.

---

I'd like to get a second approval in before we merge this.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,34 @@
+==========================================
+
+Detects if a class or struct publicly inherits from 
+``std::enable_shared_from_this``, because unintended behavior will 
```
Please synchronize with statement in Release Notes.

MichelleCDjunaidi commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
````diff
@@ -0,0 +1,34 @@
+    class BadExample : std::enable_shared_from_this<BadExample> {
+    
+    // ``shared_from_this``` unintended behaviour
+    // libstd implementation returns uninitialized ``weak_ptr``
````
oops, missed this by mistake. will fix.

MichelleCDjunaidi commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -98,6 +98,13 @@ Improvements to clang-tidy
+- New :doc:`bugprone-incorrect-enable-shared-from-this
+  <clang-tidy/checks/bugprone/incorrect-enable-shared-from-this>` check.
+
+  Detects if class/structs publicly inherits from ``std::enable_shared_from_this``,
```
Will synchronize everything.

MichelleCDjunaidi committed:
change to matcher-based check and cleanup headers

MichelleCDjunaidi committed:
synchronize check documentation wording

MichelleCDjunaidi commented:
@5chmidti @EugeneZelenko Any more feedback I need to work on?

EugeneZelenko commented:
Please resolve all fixed comments. Please rebase from `main`.

MichelleCDjunaidi commented:
...oops, I don't think that was supposed to happen.

MichelleCDjunaidi commented:
For context: force-push is to undo merge-squash. Didn't know that pinged everyone else, real sorry!

MichelleCDjunaidi commented:
...git rebased. Not sure this is the outcome we want.

EugeneZelenko commented:
Are you sure that you rebased, not merged `main`?

MichelleCDjunaidi commented:
arts@DESKTOP-1BQBBC0:~/Github/llvm-project$ git rebase origin/main
Current branch clang-tidy-create-bugprone-public-enable-shared-from-this is up to date.
arts@DESKTOP-1BQBBC0:~/Github/llvm-project$ git push
To github.com:MichelleCDjunaidi/llvm-project.git
 ! [rejected]                  clang-tidy-create-bugprone-public-enable-shared-from-this -> clang-tidy-create-bugprone-public-enable-shared-from-this (non-fast-forward)
error: failed to push some refs to 'github.com:MichelleCDjunaidi/llvm-project.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
arts@DESKTOP-1BQBBC0:~/Github/llvm-project$ git pull
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint: 
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.
arts@DESKTOP-1BQBBC0:~/Github/llvm-project$ git config pull.rebase true
arts@DESKTOP-1BQBBC0:~/Github/llvm-project$ git pull
Successfully rebased and updated refs/heads/clang-tidy-create-bugprone-public-enable-shared-from-this.
arts@DESKTOP-1BQBBC0:~/Github/llvm-project$ git push
Enumerating objects: 56952, done.
Counting objects: 100% (56951/56951), done.
Delta compression using up to 16 threads
Compressing objects: 100% (24538/24538), done.
Writing objects: 100% (47891/47891), 23.80 MiB | 2.44 MiB/s, done.
Total 47891 (delta 40331), reused 26117 (delta 23070), pack-reused 0
remote: Resolving deltas: 100% (40331/40331), completed with 7175 local objects.
To github.com:MichelleCDjunaidi/llvm-project.git
   ae9caaea3e40..3325e3be9493  clang-tidy-create-bugprone-public-enable-shared-from-this -> clang-tidy-create-bugprone-public-enable-shared-from-this

@EugeneZelenko pretty sure I did.

EugeneZelenko commented:
Looks like latest commits are from August. Did you get most recent `main` with `git fetch --prune`? Also `git push --force origin <branch name>` should be used after rebase.

MichelleCDjunaidi commented:
that's strange, I invoked git rebase with origin/main though. but will redo.

MichelleCDjunaidi commented:
@EugeneZelenko thanks for the pointers; didn't realize git fetch did not update local main. does this look fine now?

EugeneZelenko commented:
As far as I see from branch's history, it's still outdated `main` (as of September 9).

MichelleCDjunaidi commented:
Will try to resolve after lunch.

MichelleCDjunaidi commented:
Main history should be resolved now.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -131,6 +131,13 @@ New checks
   Finds cases when an uninstantiated virtual member function in a template class 
   causes cross-compiler incompatibility.
 
+- New :doc:`bugprone-incorrect-enable-shared-from-this
```
Please keep alphabetical order (by check name) in this list.

MichelleCDjunaidi committed:
sort ReleaseNotes.rst new checks by alphabetical

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,62 @@
+      // BaseSpec->getType().getAsString() results in
+      // std::enable_shared_from_this<ClassName> or
+      // alias/typedefs of std::enable_shared_from_this<ClassName>
+      "public " + BaseSpec->getType().getAsString();
```
The base type might be expanded from a macro, which we would not want to expand:

```c++
namespace std {
  template <typename T>
  struct enable_shared_from_this {};
}
struct A : std::enable_shared_from_this<A> {};
#define MACRO_A A
class B : MACRO_A {};

class B : public MACRO_A {}; // intended fix
class B : public A {}; // current fix
```

The solution would be to do a replacement on `SourceRange(Base->getBeginLoc(), Base->getBaseTypeLoc())` with only `public ` as the replacement string instead.
(Please add tests for this)

---

Note that this also has the added benefit of less potential for overlapping fixit ranges that could block each other.

MichelleCDjunaidi replied:
will implement soon when I have the bandwidth.

MichelleCDjunaidi replied:
approach confirmed to not work due to getBaseTypeLoc() not including the virtual or the access specifier. As a result, doing a blind SourceRange(BaseSpec->getBeginLoc(), BaseSpec->getBaseTypeLoc()) replacement with "public" just ends up with 

```
 //alias the template itself
 template <typename T> using esft_template = std::enable_shared_from_this<T>;
 
-class PrivateAliasTemplateClassBase : private esft_template<PrivateAliasTemplateClassBase> {};
+class PrivateAliasTemplateClassBase : public<PrivateAliasTemplateClassBase> {};
 //
 //
 
-class DefaultAliasTemplateClassBase : esft_template<DefaultAliasTemplateClassBase> {}; 
+class DefaultAliasTemplateClassBase : public<DefaultAliasTemplateClassBase> {}; 
 //
 //
 
 class PublicAliasTemplateClassBase : public esft_template<PublicAliasTemplateClassBase> {}; 
 
-struct PrivateAliasTemplateStructBase : private esft_template<PrivateAliasTemplateStructBase> {}; 
+struct PrivateAliasTemplateStructBase : public<PrivateAliasTemplateStructBase> {}; 
 //
 //

 struct A : std::enable_shared_from_this<A> {};
 #define MACRO_A A
-class B : MACRO_A {};
-class C : private MACRO_A {};
+class B : public {};
+class C : public {};
```

advice would be appreciated, because as it stands it doesn't seem possible to extract only the access specifier range. Current approach in my head is probably having a dedicated if statement just for macros, but that's kind of not ideal @5chmidti

MichelleCDjunaidi replied:
also, side note: MACRO_A doesn't count as direct inheritance either, so the fix won't emit on MACRO_A if we do the "issue FixItHint on direct inheritance only"

5chmidti replied:
> also, side note: MACRO_A doesn't count as direct inheritance either, so the fix won't emit on MACRO_A if we do the "issue FixItHint on direct inheritance only", which sort of resolves this comment indirectly as the issue is with FixItHint.

IMO macros are similar to aliases in this context, and we can do a replacement as well.

---

I found a way that works:

```diff
-  const SourceRange ReplacementRange = BaseSpec->getSourceRange();
-  const std::string ReplacementString =
-      // BaseSpec->getType().getAsString() results in
-      // std::enable_shared_from_this<ClassName> or
-      // alias/typedefs of std::enable_shared_from_this<ClassName>
-      "public " + BaseSpec->getType().getAsString();
+  const bool HasWrittenAccessSpecifier =
+      BaseSpec->getAccessSpecifierAsWritten() != AS_none;
+  const auto ReplacementRange = CharSourceRange(
+      SourceRange(BaseSpec->getBeginLoc(), BaseSpec->getBeginLoc()),
+      HasWrittenAccessSpecifier);
+  const llvm::StringRef Replacement =
+      HasWrittenAccessSpecifier ? "public" : "public ";
   const FixItHint Hint =
       IsEnableSharedFromThisDirectBase
-          ? FixItHint::CreateReplacement(ReplacementRange, ReplacementString)
+          ? FixItHint::CreateReplacement(ReplacementRange, Replacement)
```

with the following tests:
```c++
#define MACRO_PARAM(CLASS) std::enable_shared_from_this<CLASS>
class E : MACRO_PARAM(E) {};
// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: 'E' is not publicly inheriting from 'std::enable_shared_from_this', which will cause unintended behaviour when using 'shared_from_this'; make the inheritance public [bugprone-incorrect-enable-shared-from-this]
// CHECK-FIXES: class E : public MACRO_PARAM(E) {};
class F : private MACRO_PARAM(F) {};
// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: 'F' is not publicly inheriting from 'std::enable_shared_from_this', which will cause unintended behaviour when using 'shared_from_this'; make the inheritance public [bugprone-incorrect-enable-shared-from-this]
// CHECK-FIXES: class F : public MACRO_PARAM(F) {};
class G : public MACRO_PARAM(G) {};
```

MichelleCDjunaidi replied:
compiled and confirmed it works. Didn't know you can extract the SourceRange by feeding CharSourceRange getBeginLoc() twice (which is what surprised me the most, I would have thought it was getBeginLoc and getEndLoc) and the ITR boolean. Is this something that's learnable from the documentation otherwise?

5chmidti replied:
I only know of https://clang.llvm.org/docs/InternalsManual.html#sourcerange-and-charsourcerange, which is a rather short description. 
This is essentially saying: at the location (range), if there is a written access specifier, consider the range to be a token range (which will include the access specifier token in the range to be replaced), and if there is no access specifier written it is a character range (of size 0, so the replacement acts as an insertion).

---

nit: I forgot you can pass a single location to `SourceRange()`, and it will duplicate the location internally for begin and end.

MichelleCDjunaidi replied:
yes, will fix the nit and git merge origin to resolve the conflict.

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,62 @@
+      // std::enable_shared_from_this<ClassName> or
+      // alias/typedefs of std::enable_shared_from_this<ClassName>
+      "public " + BaseSpec->getType().getAsString();
+  const FixItHint Hint =
+      FixItHint::CreateReplacement(ReplacementRange, ReplacementString);
```
We should be careful with this fix. In the case of direct inheritance, it is obvious that we want to make the inheritance `public`. But in the transitive inheritance case, marking the inheritance `public` may in turn make other classes publically inherited by, which can change behavior. IMO the check should only emit the fix when `IsEnableSharedFromThisDirectBase` is `true`.

MichelleCDjunaidi replied:
fair enough, then just emit the warning and not the fix on transitive, is that what you're saying?

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace clang::tidy::bugprone {
+
+/// Detects if a class or struct publicly inherits from
+/// ``std::enable_shared_from_this``, because unintended behavior will
+/// otherwise occur when calling ``shared_from_this``.
```
This sentence should start with `Checks that` instead of `Detects if`, because *detecting* may imply that this is the bad case, when the public inheritance is the right way to do it. (Sync with the other two locations)

MichelleCDjunaidi replied:
fair enough, will do.

MichelleCDjunaidi replied:
See https://github.com/llvm/llvm-project/pull/102299#discussion_r1787090384 for reference.

5chmidti reviewed (commented):
After these, this should be good to go in my opinion.

---

For updating the PR branch:

The whole branch gets squashed when merged anyway. A merge reduces the chance for pushing already existing commits that changed during rebase, causing other people to be tagged as well. Having additional commits for things done after the review has started instead of fixing up the original commit and force-pushing has the benefit of showing the incremental progress that GitHub does not correctly show when force-pushing.

MichelleCDjunaidi commented:
@5chmidti ah, so if I'm getting you correctly, next time I shouldn't rebase and just wait to fix the conflict on merge?

5chmidti reviewed (commented):
> next time I shouldn't rebase and just wait to fix the conflict on merge

PRs that are quite outdated should, and those that have conflicts must be updated before the PR is merged. You can either rebase with upstream changes, or merge with upstream. Which one you chose is up to you, but merging seems to be less likely to cause issues with commits that are not yours from appearing in the PR. E.g., right now there is a conflict on the release notes. You could use the GitHub UI to fix the conflict, or use your local git by fetching from the remote, and either using `git rebase origin/main` or `git merge origin/main`

MichelleCDjunaidi committed:
Change fix to emit only on direct inheritance

MichelleCDjunaidi committed:
reworded first sentence of docs

MichelleCDjunaidi committed:
additional macro testing

MichelleCDjunaidi committed:
refactor add newline on eof of test file

5chmidti reviewed (approved):
Minus my comment, this looks good to me. Thanks for working on this.

MichelleCDjunaidi committed:
refactor to only replace access specifier

MichelleCDjunaidi committed:
refactor to remove duplicated arg

MichelleCDjunaidi committed:
Merge remote-tracking branch 'origin/main' into HEAD

MichelleCDjunaidi commented:
Note: I'll need someone to merge this in for me, as I don't have write access.

MichelleCDjunaidi commented:
@PiotrZSL @njames93 can I get your review on the latest state of this pull request?

5chmidti commented:
Unless someone also wants to review this PR, I'll merge this tomorrow. The unresolved review comments of the other reviewers have, FWICT, been resolved.

EugeneZelenko commented:
I think will be good idea to rebase branch again before merge and let others to see changes for some time.

MichelleCDjunaidi committed:
Merge branch 'main' into clang-tidy-create-bugprone-public-enable-shared-from-this

MichelleCDjunaidi committed:
Merge branch 'main' into clang-tidy-create-bugprone-public-enable-shared-from-this

MichelleCDjunaidi committed:
Merge remote-tracking branch 'origin/main' into clang-tidy-create-bugprone-public-enable-shared-from-this

MichelleCDjunaidi commented:
@5chmidti @EugeneZelenko Happy New Year! Is there anything left I need to do to get this merged?

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp:
```diff
@@ -0,0 +1,64 @@
+                                    hasDeclaration(EnableSharedFromThis))))))))
+          .bind("base_rec")));
+  Finder->addMatcher(
+      cxxRecordDecl(
```
Note: You may consider excluding system headers, may speed up analisys.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-enable-shared-from-this.rst:
```diff
@@ -0,0 +1,34 @@
+``std::enable_shared_from_this``, because unintended behavior will 
+otherwise occur when calling ``shared_from_this``.
+
+Consider the following code:
```
Note: this line isn't needed

MichelleCDjunaidi replied:
I think the wording/line has been suggested and approved by other reviewers, so I'll refrain from touching it.

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/incorrect-enable-shared-from-this.cpp:
```diff
@@ -0,0 +1,180 @@
+// RUN: %check_clang_tidy -std=c++11 %s bugprone-incorrect-enable-shared-from-this %t
```
Note: you could use -std=c+=11-or-later to fix problems with check in more modern C++

PiotrZSL reviewed (approved):
LGTM

MichelleCDjunaidi commented:
Working on it.

MichelleCDjunaidi committed:
exclude system header and update test flag

MichelleCDjunaidi commented:
@PiotrZSL @5chmidti I'll need help to merge this into the repo, as I don't have write permissions. Thank you so much and have a nice day!
